### ISSUE TITLE
sfdx:force:data:record:create and sfdx:force:data:record:update are trimming single quotes from values

### ISSUE THREAD

**Issue by jamesedwardsvtgov** [forcedotcom/cli#1820]
### Summary

sfdx:force:data:record:create and sfdx:force:data:record:update are trimming single quotes from values

### Steps To Reproduce:
`sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c='Won't fix'" --json`

`sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c='Won\\'t fix'" --json`

`sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c='Won''t fix'" --json`

`sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c='Won\\''t fix'" --json
`
### Expected result

The value should be uploaded with single quotes present as the value is a contraction.

### Actual result

The value has single quotes removed even when escaped.

### System Information

- Which shell/terminal are you using? (e.g. bash, zsh, powershell 5, powershell 7, cmd.exe, etc.)

git bash windows 10

see below for `sfdx version --verbose --json` 

{
  "cliVersion": "sfdx-cli/7.173.0",
  "architecture": "win32-x64",
  "nodeVersion": "node-v16.14.0",
  "pluginVersions": [
    "@oclif/plugin-autocomplete 1.3.1 (core)",
    "@oclif/plugin-commands 2.2.0 (core)",
    "@oclif/plugin-help 5.1.14 (core)",
    "@oclif/plugin-not-found 2.3.3 (core)",
    "@oclif/plugin-plugins 2.1.1 (core)",
    "@oclif/plugin-update 3.0.2 (core)",
    "@oclif/plugin-version 1.1.2 (core)",
    "@oclif/plugin-warn-if-update-available 2.0.6 (core)",
    "@oclif/plugin-which 2.1.0 (core)",
    "alias 2.1.5 (core)",
    "apex 1.3.0 (core)",
    "auth 2.2.14 (core)",
    "community 2.0.7 (core)",
    "config 1.4.21 (core)",
    "custom-metadata 2.0.1 (core)",
    "data 2.1.4 (core)",
    "generator 2.0.6 (core)",
    "info 2.1.10 (core)",
    "limits 2.0.5 (core)",
    "org 2.2.9 (core)",
    "packaging 1.9.4 (core)",
    "schema 2.1.10 (core)",
    "signups 1.2.7 (core)",
    "source 2.0.19 (core)",
    "telemetry 2.0.2 (core)",
    "templates 55.1.0 (core)",
    "trust 2.0.4 (core)",
    "user 2.1.12 (core)",
    "@salesforce/sfdx-plugin-lwc-test 1.0.1 (core)",
    "salesforce-alm 54.8.1 (core)"
  ],
  "osVersion": "Windows_NT 10.0.19044",
  "shell": "C:\\Users\\USERNAME\\AppData\\Local\\Programs\\Git\\usr\\bin\\bash.exe",
  "rootPath": "C:\\Users\\USERNAME\\AppData\\Roaming\\npm\\node_modules\\sfdx-cli"
}

**Comment by RodEsp** [comment]
Hey @jamesedwardsvtgov, I think I have a fix for this but wanted your opinion.

If you ran the following would you be ok with the value of `field__c` being `'Won't Fix'`?
```bash
sfdx force:data:record:create -s SObjectName -v "field__c='Won't fix'"
```

In order to get the value to just be `Won't Fix` you'd have to run it without the `'` surrounding it, like so:
```bash
sfdx force:data:record:create -s SObjectName -v "field__c=Won't fix"
```

**Comment by jamesedwardsvtgov** [comment]
Hi @RodEsp thank you for this comment.

The field is used in a roll-up calculation and the value that is used to calculate is managed - so the value in this case would need to be `Won't Fix` with the apostrophe included. 

Unfortunately as shown in the original issue summary, both examples are not working as expected and the single quote is removed.

I do see there is a PR open for this and we'll look forward to a resolution. 

Thank you and this is greatly appreciated!

**Comment by RodEsp** [comment]
Hi @jamesedwardsvtgov, just wanted to let you know that we've merged the fix for this.

You'll be able to have values with single quotes by running this:
```bash
sfdx force:data:record:create -s SObjectName -v "field__c=Won't fix"
```
or similarly for update:
```bash
sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c=Won't fix"
```
*Note that* `Won't Fix` *doesn't have any single quotes around it.*

The fix will be included in `sfdx` version `1.181.0` which will be available tomorrow as a release candidate, and be promoted to a stable release on Thursday, Dec 15th in our stable channel.

If you installed the CLI with our installers or tarballs and you want to update to the release candidate run `sfdx update stable-rc` (run `sfdx update stable` to return to the stable version).

If you have an npm installation run `npm install sfdx-cli@latest-rc -g` to update to the release candidate (run `npm install sfdx-cli -g` to return to the stable version).

<strike>I'll close this issue on Dec 15th.</strike>

**Comment by jamesedwardsvtgov** [comment]
Hi @RodEsp 

Thank you for these updates! I have installed the latest-rc with the command you provided `npm install sfdx-cli@latest-rc -g`

With the latest release candidate apostrophes are no long removed from the value, which is solving the original issue! 

**Unfortunately, there is a new issue with uploading a value that contains a space.** 

For example: 

`sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c=this contains a space"`

and 

`sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c='this contains a space'"`


is showing the error message:

`Updating Record... Failed`
`ERROR running force:data:record:update:  Malformed key=value pair for value: contains`

The value that our use case requires is `Won't fix`, which besides containing an apostrophe contains a space. 

I've included the output for `sfdx version --verbose --json`. Please let me know if any other information would be helpful at this time or steps needed on my part to move towards resolution. 
 
 Here is the sfdx version output for latest install
```json
{
  "cliVersion": "sfdx-cli/7.181.0",
  "architecture": "win32-x64",
  "nodeVersion": "node-v16.14.0",
  "pluginVersions": [
    "@oclif/plugin-autocomplete 1.3.6 (core)",
    "@oclif/plugin-commands 2.2.1 (core)",
    "@oclif/plugin-help 5.1.19 (core)",
    "@oclif/plugin-not-found 2.3.9 (core)",
    "@oclif/plugin-plugins 2.1.7 (core)",
    "@oclif/plugin-search 0.0.4 (core)",
    "@oclif/plugin-update 3.0.7 (core)",
    "@oclif/plugin-version 1.1.3 (core)",
    "@oclif/plugin-warn-if-update-available 2.0.14 (core)",
    "@oclif/plugin-which 2.2.2 (core)",
    "alias 2.1.11 (core)",
    "apex 1.3.0 (core)",
    "auth 2.3.1 (core)",
    "community 2.0.18 (core)",
    "config 1.4.22 (core)",
    "custom-metadata 2.0.5 (core)",
    "data 2.1.15 (core)",
    "generator 2.0.12 (core)",
    "info 2.2.11 (core)",
    "limits 2.1.3 (core)",
    "org 2.2.17 (core)",
    "packaging 1.10.10 (core)",
    "schema 2.1.18 (core)",
    "signups 1.2.12 (core)",
    "source 2.3.7 (core)",
    "telemetry 2.0.5 (core)",
    "templates 55.1.0 (core)",
    "trust 2.2.1 (core)",
    "user 2.1.22 (core)",
    "@salesforce/sfdx-plugin-lwc-test 1.0.1 (core)",
    "salesforce-alm 54.8.4 (core)"
  ],
  "osVersion": "Windows_NT 10.0.19044",
  "shell": "C:\\Users\\user.name\\AppData\\Local\\Programs\\Git\\usr\\bin\\bash.exe",
  "rootPath": "C:\\Users\\user.name\\AppData\\Roaming\\npm\\node_modules\\sfdx-cli"
}
```

**Comment by RodEsp** [comment]
Hey @jamesedwardsvtgov, if you want to include values that contain a space you can surround them by single quotes, like so: 
```bash
sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c='this contains a space'"
```

You're right that there's still an issue with values that contain both a `'` _and_ a space though 🤔
We've been struggling to remedy that use case while maintaining backward compatibility with others, such as including Unicode characters.

I'll give this another try in the next few days. I've already re-opened our internal bug record.

**Comment by jamesedwardsvtgov** [comment]
Thank you I will look forward to the resolution of this issue to use the CLI. 

I edited my previous comment and want to clarify that both: 

`sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c=this contains a space"`

and

`sfdx force:data:record:update -u sfUserName -s SObjectName -i SObjectID -v "field__c='this contains a space'"`

result in the same error which is: 

`Updating Record... Failed`
`ERROR running force:data:record:update: Malformed key=value pair for value: contains`

Thanks again for your assistance with this, and I might suggest to include in the documentation that a value containing an apostrophe does not currently work on the stable branch.


In the meantime is there a suggested workaround for this to update or create record values with spaces and apostrophes using the CLI?

**Comment by RodEsp** [comment]
Hey @jamesedwardsvtgov, that's very strange 🤔
The spaces work for us no problem.

This is the command running with the same version of the CLI as you
![image](https://user-images.githubusercontent.com/1084688/206569096-d4c6e105-93dd-4408-aa95-f96d74d1382f.png)

And the value pulled up in the developer console using a SOQL query
![image](https://user-images.githubusercontent.com/1084688/206569213-402d7103-f2fa-4679-91c4-d4430657f75b.png)

We will continue to look into this though.
I'm not sure about a workaround for now using the CLI but also looking into it.

**Comment by RodEsp** [comment]
For a workaround you could use the `sfdx force:data:bulk:upsert` command, it's definitely overkill for just updating one value but it could let you move forward for now.

Checkout `sfdx force:data:bulk:upsert --help` for more info. In short, you'll want to create a CSV file that contains the field names and values you want to update for the record type. You can learn more about the format for that CSV file in our Bulk API 2.0 docs: [Prepare CSV Files](https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/datafiles_prepare_csv.htm).

**Comment by jamesedwardsvtgov** [comment]
@RodEsp I am sorry yes that is working to save values with spaces!  

You are correct it is the combination of apostrophe and space that is causing the issue. I see that you have included this info in [1837](https://github.com/forcedotcom/cli/issues/1837) and I will monitor that issue for updates. 

Thank you also for the suggestion to use the `sfdx force:data:bulk:upsert`. We will try this approach.

**Comment by RodEsp** [comment]
Hi @jamesedwardsvtgov, I've been doing some testing and the fix implemented in https://github.com/salesforcecli/plugin-data/pull/427 does allow for single quotes inside values with spaces, but there's some weird interplay in how different shells parse strings and how the CLI parses flag values that create some weird behavior.

I'm going to keep testing and see if we can at least update our docs and `--help` output to make that interplay easier to understand.

For now here's an example of how you can add your `Won't Fix` value using both Powershell and GitBash:

* GitBash
    * Enclose the value of your `key=value` pair in `"` and use `\` to escape them.
![image](https://user-images.githubusercontent.com/1084688/207157774-6bfb21d2-78ea-44bc-bbdc-909349798f93.png)

* Powershell
    * Enclose the value of your `key=value` pair in `"` and use `` \` `` to escape them.
     (I'm not entirely sure why `` \` `` is needed in that order but `` ` `` is the escape character for Powershell and I'm assuming `\` does something in the CLI's flag parsing logic to escape the `` ` ``).
![image](https://user-images.githubusercontent.com/1084688/207158688-8bba5ada-a5e0-4703-af83-c4c883a653be.png)

**Comment by jamesedwardsvtgov** [comment]
@RodEsp thank you for this update! Your suggested work around to use the bulk upsert command has worked great for our use case and we are able to proceed.

**Comment by iowillhoit** [comment]
👋 This has been released in today's latest build 7.181.1
